cmd/ketch/app_log.go

```diff
@@ -0,0 +1,364 @@
+	}
+	set := map[string]string{
+		ketchAppNameLabel: options.appName,
+	}
+	if len(options.processName) > 0 {
+		set[ketchProcessNameLabel] = options.processName
+	}
+	if options.deploymentVersion > 0 {
+		set[ketchDeploymentVersionLabel] = fmt.Sprintf("%d", options.deploymentVersion)
+	}
+	s := labels.SelectorFromSet(set)
+	opts := watchOptions{
+		namespace:    pool.Spec.NamespaceName,
+		selector:     s,
+		follow:       options.follow,
+		ignoreErrors: options.ignoreErrors,
+		timestamps:   options.timestamps,
+		prefix:       options.prefix,
+		out:          out,
+	}
+	return watchLogs(cfg.KubernetesClient(), opts, readLogs, streamLogs)
+}
+
+type watchOptions struct {
+	namespace    string
+	selector     labels.Selector
+	follow       bool
+	ignoreErrors bool
+	timestamps   bool
+	prefix       bool
+	out          io.Writer
+}
+
+// ketchContainerName returns a name of an application container.
+// A pod can have several containers, one of them is defined and created by ketch, it's an application container.
+// The others can be injected by istio, vault, etc.
+func ketchContainerName(pod corev1.Pod) (*string, error) {
+	// this is an application pod.
+	// it has one app container, we don't care about the other containers.
+	for _, c := range pod.Spec.Containers {
+		if strings.HasPrefix(pod.Name, c.Name) {
+			return &c.Name, nil
+		}
+	}
+	return nil, fmt.Errorf("pod %s doesn't have an app container", pod.Name)
+}
+
+func isContainerRunning(pod corev1.Pod, containerName string) bool {
+	for _, container := range pod.Status.ContainerStatuses {
+		if container.Name == containerName {
+			return container.State.Running != nil
+		}
+	}
+	return false
+}
+
+type readLogsFn func(getLogs getLogsFn, pod corev1.Pod, containerName string, out io.Writer) chan logMessage
+type streamLogsFn func(getLogs getLogsFn, pod corev1.Pod, containerName string, out io.Writer, lastTime time.Time, msgCh chan logMessage) chan struct{}
+
+func watchLogs(cli kubernetes.Interface, options watchOptions, readLogs readLogsFn, streamLogs streamLogsFn) error {
+	pods, err := cli.CoreV1().Pods(options.namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: options.selector.String()})
+	if err != nil {
+		return err
+	}
+	// we are going to read logs from all running pods, just read without streaming
+	msgChs := make(map[types.UID]chan logMessage, len(pods.Items))
+	for _, pod := range pods.Items {
+		containerName, err := ketchContainerName(pod)
+		if err != nil {
+			return err
+		}
+		msgChs[pod.UID] = readLogs(cli.CoreV1().Pods(pod.Namespace).GetLogs, pod, *containerName, options.out)
+	}
+
+	// we want to show the logs sorted by timestamp.
+	// lets store one message per pod and then in a loop below we will select one message with minimal time on each iteration.
+	messages := make(map[types.UID]logMessage, len(msgChs))
+	for podUID, msg := range msgChs {
+		if m, ok := <-msg; ok {
+			messages[podUID] = m
```

Comment:
not sure but, could we set it in the loop above?
https://github.com/shipa-corp/ketch/pull/45/files#diff-80131f1268c6db0851cfbef6ae1ef9c7d4b8f7e5ed98ffac81ffd8cf7c47707fR144

Reply:
yep, though it's possible I would keep it as is. The first loop runs several goroutines, each of them opens a connection to read logs. This gives a small performance boost)